JITMs: add component to be used in the Jetpack Dashboard.

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

JITMs: add component to be used in the Jetpack Dashboard.

<img width="868" alt="screenshot 2018-12-05 at 17 17 41" src="https://user-images.githubusercontent.com/426388/49527273-db7b7080-f8b1-11e8-8b53-f3cd6a41cc2a.png">
<img width="835" alt="screenshot 2018-12-05 at 17 17 29" src="https://user-images.githubusercontent.com/426388/49527275-dc140700-f8b1-11e8-8ae9-c16768431556.png">
<img width="828" alt="screenshot 2018-12-05 at 17 19 12" src="https://user-images.githubusercontent.com/426388/49527297-e7673280-f8b1-11e8-98af-669bd12946d6.png">
<img width="897" alt="screenshot 2018-12-06 at 11 55 06" src="https://user-images.githubusercontent.com/426388/49579890-d3710e80-f94d-11e8-9a14-160a06a06bd2.png">
<img src="https://user-images.githubusercontent.com/426388/49581453-509e8280-f952-11e8-9561-783cad310e91.gif">

The first part of this PR (the necessary new method in the rest api client and the new redux structure) was merged in #10818. This PR brings the second part: a new Jitm component and a new QueryJitm component, as well as a call to the new Jitm component in the main dashboard.

- [x] Check all the to-dos for things to do. 
- [x] Dequeue regular JITMs on the JP dashboard pages.
- [x] Figure out what to do with existing JITMs; with the current implementation we have no JITMs that are set to be displayed on JP dash. We will most likely need to edit the JITMs on wpcom to add the string matching JP Dash for some messages.
- [x] Some HTML encoding issues with the messages.

#### Testing instructions:

* You'll need a WordPress.com sandbox.
* Start by applying D19944-code.
* Sandbox your Jetpack site with the `JETPACK__SANDBOX_DOMAIN` constant ([reference](https://github.com/Automattic/jetpack/blob/add/jitm-component/docker/README.md#must-use-plugins-directory)). 
    - If you don't have a WordPress.com sandbox, you can sandbox mine. Through a Jurassic Ninja site with that branch, go to Settings > Jetpack Constants, and add `sandboxme.wordpress.com` as the sandbox you want to use.
* Visit Jetpack > Dashboard: a message should appear.
* Visit Jetpack > Settings: the message should disappear.
* Visit Jetpack > Plans: a different message should appear.

#### Known issues / To do:

- `toplevel_page_jetpack` should continue to work as a string one can use to target **all** pages of the Jetpack dashboard at once.
- ~Lists are not displayed properly right now.~
- ~Props warnings for `isActivatingModule` and `isModuleActivated`~
- ~When dismissing a message, and then switching to a different tab, the dismissed message is visible for a few seconds until the new message (or no message at all) is loaded.~
- ~When activating a module via a JITM, the JITM remains in place once the module has been activated. This is related to #9462~ I opted to remove the button once the module has been activated. This way one can still click on the CTA to learn more if they are interested. If they leave the page or refresh, the JITM will disappear for good.
- ~The look of the messages is a bit different from the ones you'll find in other places in wp-admin. The CTA button uses Calypso's button component, and as such the text inside the button is in uppercase. That is not the case with the wp-admin implementation of the button.~

#### Proposed changelog entry for your changes:

* Notices: display more relevant notices depending on the Jetpack dashboard screen you visit.

jeherve committed:
JITM: first pass at a query component to gather JITMs in JP dash.

jeherve committed:
Pass message path when fetching a JITM

jeherve committed:
Temporary hardcoded message path

There are no dash-specific JITMs yet

jeherve committed:
Add component contents: message body, event handlers

oskosk commented on _inc/client/components/jitm/index.jsx:
```diff
@@ -0,0 +1,193 @@
+
+		const cta = get( jitm, 'CTA', null );
+		const ctaMessage = get( cta, 'message', null );
+		const ctaNewWindow = get( cta, 'newWindow', null );
```
`ctaNewWindow` seems to be be unused.

oskosk replied:
Fixed in #eaf26aaaf0613a8ae57692831f4b6ff4d30af416

oskosk commented on _inc/client/components/jitm/index.jsx:
```diff
@@ -0,0 +1,193 @@
+								<div className="jitm-banner__action">
+									<Button
+										// to-do: missing the option to open in a new window with the ctaNewWindow const.
+										// target={ ctaNewWindow === false ? '_self' : '_blank' }
```
ah here's the usage of ctaNewWindow, commented out

oskosk commented:
`_inc/client/components/jitm/style.scss` is empty right now

oskosk commented on _inc/client/state/reducer.js:
```diff
@@ -39,7 +40,8 @@ const jetpackReducer = combineReducers( {
 	publicize,
-	siteVerify
+	siteVerify,
+	jitm
```
A comma here is ok if you want it.

oskosk commented:
`_inc/client/components/jitm/style.scss` is empty right now

oskosk commented:
I'm seeing the message encoded. Maybe we can use `lib/decode-entities/`.

![image](https://user-images.githubusercontent.com/746152/49314820-5b789380-f4ca-11e8-82b2-f5f5c729ed3c.png)

oskosk commented on _inc/client/components/jitm/index.jsx:
```diff
@@ -0,0 +1,193 @@
+										className="jitm-button"
+										primary={ activate_module === null && ctaPrimary ? true : false }
+										compact={ true }
+										onClick={ this.trackLearnMore( feature_class ) }s
```
Tipo here, an extrac `s` at the end of the line.

Symptom was that `Button` gets rendered with and `s` prop.

```
warning.js:33 Warning: Unknown prop `s` on <a> tag. Remove this prop from the element. For details, see https://fb.me/react-unknown-prop
    in a (created by Button)
    in Button (created by Jitm)
    in div (created by Jitm)
```

oskosk replied:
Fixed in 63ad0ec97e39ed443462fa53383fae2d9d46e0f5

oskosk committed:
Remove unused character

oskosk commented on _inc/client/components/jitm/index.jsx:
```diff
@@ -0,0 +1,193 @@
+	};
+
+	renderListItem = ( list, id, feature_class ) => {
+		for ( const i of list ) {
```
let's use a more significant name here for `i`. even when it's straightforward, we should use `i` just on `for` loops, and/or standing for `index`.

oskosk commented:
We can split this work in two PRs at least. This is how we've done a few things...

* One for Redux-related stuff (action, reducer selectors).
* Another one for introducing the component.

It would be cool to at least provide one or two unit tests for the redux piece of code. they're straightforward if you take another one as example. This usually speeds up work because we can easily merge the redux work if the tests pass.

oskosk reviewed (changes requested):
I think I commented all of the previous things without using a Review.

jeherve committed:
Prettier: apply to new file

jeherve committed:
Use a more descriptive const name.

@see https://github.com/Automattic/jetpack/pull/10759#discussion_r238007121

jeherve committed:
Use decodeEntities to avoid HTML encoding issues.

jeherve committed:
Update tracking for all actions.

oskosk reviewed (dismissed):
LGTM!

jeherve committed:
Add JITM dismissal functionality

jeherve committed:
Import existing SCSS

jeherve committed:
Do not enqueue default JITMS in the Jetpack dashboard.

jeherve committed:
Track JITM dismissal.

jeherve committed:
Fetch a new JITM when we move to a different subpage.

Each subpage has the following format:
toplevel_page_jetpack_section
where section matches the subpage's path.

jeherve committed:
Update tracking: remove tracking that is already done server-side

jeherve committed:
Update name of selector to avoid confusion

jeherve committed:
Switch to a element for link instead of Button

This allows me to add the "target" we need

jeherve committed:
Remove comments

jeherve committed:
Add extra &redirect parameter just like in regular JITMs

We can hardcode this one as we are always on the same page.

jeherve committed:
Add unique key as props to JITM component.

This allows to force the component to render again every time props change.
In this case, we force the component to render again every time we load a different subpage of the Jetpack
dashboard.

jeherve committed:
Remove unneeded SCSS file

We load our styles from the existing JITM styles instead.

jeherve committed:
Handle special case of Jetpack > Settings > Writing

When someone navigates from Settings > {something} to Settings > Writing, we want to display messages that
are targetting the overall Settings page.

jeherve commented:
@AnnaMag @keoshi I'd be happy if you could give this a try when you get the chance! Thank you!

jeherve committed:
Don't attempt to display the message when it's still being fetched

jeherve committed:
Ensure CTA buttons are not all uppercase.

jeherve committed:
Hide the activate button once a module has been activated.

oskosk committed:
Use isModuleActivated and isActivatingModule as boolean props instead of functions

oskosk committed:
Remove unnecessary aliasing of isModuleActivated import

This is only needed if one wants to refer to a prop after destructuring instead of referring to it as this.props.isModuleActivated, so to avoid collision of name scopes

oskosk committed:
Use module_slug instead of activate_module as var name for readability

jeherve committed:
Fix rendering of list items.

jeherve committed:
Prettier formatting

jeherve committed:
Avoid trailing 0 when we have no list.

oskosk reviewed (approved):
LGTM!

dereksmart commented on _inc/client/components/jitm/index.jsx:
```diff
@@ -0,0 +1,228 @@
+import { isModuleActivated, activateModule, isActivatingModule } from 'state/modules';
+import { getJitmDismissalResponse, getJitm, isDismissingJitm, isFetchingJitm } from 'state/jitm';
+
+require( './../../../../scss/jetpack-admin-jitm.scss' );
```
What is this for?

jeherve commented:
Closing this in favor of #10889

jeherve commented:
Previous changes will be reverted in #10890
